cpp/src/reductions/simple.cuh

```diff
@@ -280,7 +280,7 @@ struct same_element_type_dispatcher {
   template <typename ElementType>
   static constexpr bool is_supported()
   {
-    return !cudf::is_dictionary<ElementType>();
+    return !cudf::is_dictionary<ElementType>() && not std::is_same_v<ElementType, void>;
```

Comment:
How would `void` happen?

Reply:
Yeah this is interesting. We need to disable the code path for non-integral types. I tried to add this:
```
template <type_id t>
struct dispatch_void_if_non_integral {
  using type = std::conditional_t<std::is_integral_v<id_to_type<t>>, id_to_type<t>, void>;
};
```
and ran into a compiler error when this still invokes the function `same_element_type_dispatcher::operator<void>()`.

Reply:
Ok. That is fine.

Reply:
It is a bit weird to mix `!` and `not` in the same line... Let's try to use `!/&&` or `not/and` consistently?